torch/distributed/pipelining/schedules.py

```diff
@@ -612,6 +648,78 @@ def __init__(
             if self._has_backward:
                 stage._prepare_backward_infra(n_microbatches)
 
+    def dump_csv(self, filename):
+        with open(filename, "w", newline="") as csvfile:
+            writer = csv.writer(csvfile)
+            for rank in self.pipeline_order:
+                writer.writerow(self.pipeline_order[rank])
+
+    def _validate_schedule(self):
+        def _validate_rank_actions(
```

Comment:
This is pretty good to validate for a single rank, but https://github.com/pytorch/pytorch/blob/main/test/distributed/pipelining/test_schedule.py#L453-L554 tries to do it across all ranks (its pretty ugly though) and also doesn't include W.

So we can probably try to consolidate these together later

Reply:
ah, bummer i forgot about that.  lets consolidate.